### DISCUSSION TITLE
try_get_with_by_ref always triggers init future

### DISCUSSION THREAD

**Discussion by dannygale** [moka-rs/moka#206] [Q&A] [answered]
I'm having some trouble with `try_get_with_by_ref` and similar functions, and i'm not sure if it's something i'm doing wrong, or if I've found a bug. The issue is that `try_get_with_by_ref` always triggers the init future, even when `contains_key` returns true in the init future itself. To debug this issue, I'm running on a single-threaded tokio runtime

`self.mdcache` is a `moka::future::Cache<PathBuf, MyStruct>`

This code:
```rust
        let key: PathBuf = self.path.clone();
        let metadata = self
            .mdcache
            .try_get_with_by_ref(
                &key,
                async {
                    tracing::debug!(
                        "mdcache miss ({})",
                        if self.mdcache.contains_key(&key) {
                            "NOT expected"
                        } else {
                            "expected"
                        },
                    );
                    self.file.read_metadata()
                }
                .await,
            )
            .await
            .unwrap();
```
Is producing this kind of output consistently -- the only "expected" miss is the first access. For all of the following accesses, `contains_key` is true, but the init function is getting called and awaited anyway, defeating the purpose of the cache. These are consecutive accesses for the same key, produced by the above snippet:
```
2022-12-06T17:52:33.327125Z DEBUG mdcache miss (expected)
2022-12-06T17:52:33.330107Z DEBUG mdcache miss (NOT expected)
2022-12-06T17:52:33.332269Z DEBUG mdcache miss (NOT expected) 
...
```

The following code works as expected, but loses the guarantee of a single in-flight init task for a given key:
```rust
        match self.mdcache.get(&self.path) {
            Some(md) => Ok(md),
            None => {
                tracing::debug!("mdcache miss");
                let md = self.file.read_metadata().await?;
                self.mdcache.insert(self.path.clone(), md.clone()).await;

                Ok(md)
            }
        }
```

Am I misunderstanding how `try_get_with` and friends are supposed to work, am I misusing it, or is this an issue with the cache implementation?

**Comment by dannygale** [comment]
Pretty sure this is a cache implementation issue. Created https://github.com/moka-rs/moka/issues/208

**Reply by tatsuya6502** [reply]
Thank you for filing #208. I reviewed but found a problem in the reproducing code. I explained how to fix it. I closed the issue as I believe `try_get_with_by_ref` is working correctly. Please reopen it if needed.

**Comment by tatsuya6502** [comment, marked as answer]
Thank you for reporting. I reviewed your original code, and I believe you have an extra `.await`, which causes the async block executed unconditionally _before_ being passed to `try_get_with_by_ref()`. Please remove it.

```rust
        let metadata = self
            .mdcache
            .try_get_with_by_ref(
                &key,
                async {
                    tracing::debug!(
                        "mdcache miss ({})",
                        if self.mdcache.contains_key(&key) {
                            "NOT expected"
                        } else {
                            "expected"
                        },
                    );
                    self.file.read_metadata()
                }
                // Please remove this await. This causes the async block above executed
                // unconditionally _before_ being passed to try_get_with_by_ref().
                .await, 
            )
            .await
            .unwrap();
```

**Reply by tatsuya6502** [reply]
I am marking this Q&A as answered. Please let me know if you need any further assistance.
